File: backend/controllers/application_controller.py
from flask import request, jsonify
from models.application import create_application, get_applications_by_user, get_applications_by_job, update_application_status

def apply_for_job(current_user, job_id):
    if current_user['role'] != 'seeker':
        return jsonify({'message': 'Only seekers can apply for jobs'}), 403
        
    data = request.get_json() or {}
    try:
        app_id = create_application(
            job_id=job_id,
            user_id=current_user['user_id'],
            resume_link=data.get('resume_link')
        )
        return jsonify({'message': 'Application submitted successfully', 'application_id': app_id}), 201
    except Exception as e:
        return jsonify({'message': str(e)}), 500

def list_user_applications(current_user):
    try:
        applications = get_applications_by_user(current_user['user_id'])
        return jsonify({'applications': applications}), 200
    except Exception as e:
        return jsonify({'message': str(e)}), 500

def list_job_applications(current_user, job_id):
    try:
        applications = get_applications_by_job(job_id, current_user['user_id'])
        return jsonify({'applications': applications}), 200
    except Exception as e:
        return jsonify({'message': str(e)}), 500

def change_application_status(current_user, application_id):
    data = request.get_json()
    if not data or not data.get('status'):
        return jsonify({'message': 'Status is required'}), 400
        
    valid_statuses = ['pending', 'reviewed', 'accepted', 'rejected']
    if data['status'] not in valid_statuses:
        return jsonify({'message': 'Invalid status'}), 400
        
    try:
        success = update_application_status(
            application_id=application_id,
            employer_id=current_user['user_id'],
            status=data['status']
        )
        if success:
            return jsonify({'message': 'Application status updated successfully'}), 200
        return jsonify({'message': 'Application not found or unauthorized'}), 404
    except Exception as e:
        return jsonify({'message': str(e)}), 500

File: backend/controllers/applicationController.js
const { createApplication, getApplicationsByUser, getApplicationsByJob, updateApplicationStatus } = require('../models/applicationModel');

const applyForJob = async (req, res) => {
  try {
    if (req.user.role !== 'seeker') {
      return res.status(403).json({ message: 'Only seekers can apply for jobs' });
    }

    const jobId = req.params.job_id;
    const userId = req.user.user_id;
    const { resume_link } = req.body || {};

    const appId = await createApplication(jobId, userId, resume_link);
    res.status(201).json({ message: 'Application submitted successfully', application_id: appId });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

const listUserApplications = async (req, res) => {
  try {
    const userId = req.user.user_id;
    const applications = await getApplicationsByUser(userId);
    res.status(200).json({ applications });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

const listJobApplications = async (req, res) => {
  try {
    const jobId = req.params.job_id;
    const employerId = req.user.user_id;
    
    const applications = await getApplicationsByJob(jobId, employerId);
    res.status(200).json({ applications });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

const changeApplicationStatus = async (req, res) => {
  try {
    const applicationId = req.params.application_id;
    const employerId = req.user.user_id;
    const { status } = req.body;

    if (!status) {
      return res.status(400).json({ message: 'Status is required' });
    }

    const validStatuses = ['pending', 'reviewed', 'accepted', 'rejected'];
    if (!validStatuses.includes(status)) {
      return res.status(400).json({ message: 'Invalid status' });
    }

    const success = await updateApplicationStatus(applicationId, employerId, status);
    if (success) {
      return res.status(200).json({ message: 'Application status updated successfully' });
    }
    res.status(404).json({ message: 'Application not found or unauthorized' });
  } catch (err) {
    res.status(500).json({ message: err.message });
  }
};

module.exports = {
  applyForJob,
  listUserApplications,
  listJobApplications,
  changeApplicationStatus
};
